rust/kernel/debugfs.rs

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust implementation of `debugfs`.
+//!
+//! This module allows Rust kernel modules to create directories and files in
+//! `/debugfs`.
+//!
+//! C header: [`include/linux/debugfs.h`](../../../include/linux/debugfs.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/debugfs.html>
+
+use alloc::{boxed::Box, vec::Vec};
+use core::{
+    marker::{PhantomData, Sync},
+    ptr,
+};
+
+use crate::{bindings, c_types, error, str::CStr, types::PointerWrapper, Result};
+
+/// An `inode` for a directory in debugfs.
+pub struct DebugFsDirectory {
+    dentry: *mut bindings::dentry,
+    has_parent: bool,
+    file_children: Vec<Box<dyn Sync>>,
```

Comment:
Is there anything we can do with [`to_raw_parts`](https://doc.rust-lang.org/stable/core/primitive.pointer.html#method.to_raw_parts) that would obviate the need to box the file entries?

We already depend on `ptr_metadata`.